File: v1/src/main/java/com/api/v1/medical_slots/MedicalSlot.java
package com.api.v1.medical_slots;

import com.api.v1.doctors.Doctor;
import com.api.v1.medical_slots.enums.MedicalSlotStatus;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;
import java.util.UUID;

@Document(collection = "MedicalSlots")
public record MedicalSlot(

    @Id
    String id,
    Doctor doctor,
    MedicalSlotStatus status,
    LocalDateTime availableAt,
    LocalDateTime createdAt,
    LocalDateTime cancelledAt,
    LocalDateTime completedAt
) {
    public static MedicalSlot of(Doctor doctor, LocalDateTime availableAt) {
        return new MedicalSlot(
                UUID.randomUUID().toString(),
                doctor,
                MedicalSlotStatus.ACTIVE,
                availableAt,
                LocalDateTime.now(),
                null,
                null
        );
    }

    public MedicalSlot markAsCancelled() {
        return new MedicalSlot(
                this.id,
                this.doctor,
                MedicalSlotStatus.CANCELLED,
                this.availableAt,
                this.createdAt,
                LocalDateTime.now(),
                null
        );
    }

    public MedicalSlot markAsCompleted() {
        return new MedicalSlot(
                this.id,
                this.doctor,
                MedicalSlotStatus.COMPLETED,
                this.availableAt,
                this.createdAt,
                null,
                LocalDateTime.now()
        );
    }

    public MedicalSlotResponseDTO toDTO() {
        if (status.equals(MedicalSlotStatus.CANCELLED)) {
            CancelledMedicalSlotResponseDTO.from(this);
        }
        else if (status.equals(MedicalSlotStatus.COMPLETED)) {
            CompletedMedicalSlotResponseDTO.from(this);
        }
        return MedicalSlotResponseDTO.from(this);
    }
}

File: v2/src/main/kotlin/com/api/v2/medical_slots/MedicalSlot.kt
package com.api.v2.medical_slots

import com.api.v2.doctors.Doctor
import com.api.v2.medical_slots.enums.MedicalSlotStatus
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.Document
import java.time.LocalDateTime
import java.util.UUID

@Document(collection = "MedicalSlots")
data class MedicalSlot(
    @Id
    val id: String = UUID.randomUUID().toString(),
    val doctor: Doctor,
    val availableAt: LocalDateTime,
    val status: MedicalSlotStatus,
    val createdAt: LocalDateTime,
    val cancelledAt: LocalDateTime?,
    val completedAt: LocalDateTime?
) {

    companion object {
        fun of(doctor: Doctor, availableAt: LocalDateTime): MedicalSlot {
            return MedicalSlot(
                UUID.randomUUID().toString(),
                doctor,
                availableAt,
                MedicalSlotStatus.ACTIVE,
                LocalDateTime.now(),
                null,
                null
            )
        }
    }

    fun markAsCancelled(): MedicalSlot {
        return MedicalSlot(
            this.id,
            this.doctor,
            this.availableAt,
            MedicalSlotStatus.CANCELLED,
            this.createdAt,
            LocalDateTime.now(),
            null
        )
    }

    fun markAsCompleted(): MedicalSlot {
        return MedicalSlot(
            this.id,
            this.doctor,
            this.availableAt,
            MedicalSlotStatus.COMPLETED,
            this.createdAt,
            null,
            LocalDateTime.now()
        )
    }
}